src/layers/regularizers/instance_norm.cpp

```diff
@@ -0,0 +1,335 @@
+              + dvar * (x - mean) * 2 / (channel_size - 1));
+      }
+    }
+  }
+
+}
+
+} // namespace <anon>
+
+template <typename TensorDataType, data_layout Layout, El::Device Device>
+void instance_norm_layer<TensorDataType,Layout,Device>::bp_compute()
+{
+  const El::Int num_channels = this->get_output_dims().front();
+  const El::Int channel_size = this->get_output_size() / num_channels;
+  bp_impl(*this->get_comm(),
+          num_channels,
+          channel_size,
+          this->m_epsilon,
+          this->get_prev_activations(),
+          this->get_prev_error_signals(),
+          this->get_error_signals(),
+          this->m_workspace);
+}
+
+// =============================================
+// Builder function
+// =============================================
+
+namespace
+{
+
+template <typename T, data_layout L, El::Device D>
+struct Builder
+{
+  static std::unique_ptr<Layer> Build(...)
+  {
+    LBANN_ERROR(
+      "Attempted to construct instance_norm_layer ",
+      "with invalid parameters ",
+      "(TensorDataType=",TypeName<T>(),", ",
+      "Layout=",to_string(L),", ",
+      "Device=",to_string(D),")");
+    return nullptr;
+  }
+};
+
+template <El::Device Device>
+struct Builder<float,data_layout::DATA_PARALLEL,Device>
+{
+  template <typename... Args>
+  static std::unique_ptr<Layer> Build(Args&&... args)
+  {
+    using LayerType = instance_norm_layer<float,
+                                          data_layout::DATA_PARALLEL,
+                                          Device>;
+    return make_unique<LayerType>(std::forward<Args>(args)...);
+  }
+};
+
+template <El::Device Device>
+struct Builder<double,data_layout::DATA_PARALLEL,Device>
+{
+  template <typename... Args>
+  static std::unique_ptr<Layer> Build(Args&&... args)
+  {
+    using LayerType = instance_norm_layer<double,
+                                          data_layout::DATA_PARALLEL,
+                                          Device>;
+    return make_unique<LayerType>(std::forward<Args>(args)...);
+  }
+};
+
+} // namespace <anon>
+
+template <typename TensorDataType, data_layout Layout, El::Device Device>
+std::unique_ptr<Layer> build_instance_norm_layer_from_pbuf(
+  lbann_comm* comm, lbann_data::Layer const&)
+{
+  using BuilderType = Builder<TensorDataType, Layout, Device>;
+  return BuilderType::Build(comm);
+}
+
+// =============================================
+// Explicit template instantiation
+// =============================================
+
+#define PROTO(T)                                        \
+  template class instance_norm_layer<                   \
+    T, data_layout::DATA_PARALLEL, El::Device::CPU>
+#include "lbann/macros/instantiate.hpp"
+#undef PROTO
+
+#ifdef LBANN_HAS_GPU
+#define PROTO(T)                                        \
+  extern template class instance_norm_layer<            \
+    T, data_layout::DATA_PARALLEL, El::Device::GPU>
+#include "lbann/macros/instantiate.hpp"
+#undef PROTO
+#endif // LBANN_HAS_GPU
```

Comment:
This is fine and clearly it works. However, we put a fair bit of effort to remove as much of the explicit device permutations (things like `#ifdef LBANN_HAS_THIS_OR_THAT_DEVICE` from the developer-facing ETI garbage, and this addition is explicitly adding that back in. This is why I introduced the `*_builder.cpp` files to explicitly handle cases in which the layers require separate instantiation (e.g., because they contain CUDA device code) but their builder functions are not device-specific. There is no explicit special device handling, and the `instantiate_device.hpp` macro include implies that the `PROTO_DEVICE` spec is valid across all known device types (as it does here as well -- the main differentiating point is the first clause). @bvanessen do you have an opinion?